[custom channels 5/5]: merge custom channel staging branch into master

guggero opened the pull request:
Fixes https://github.com/lightningnetwork/lnd/issues/8769.

This is the cleaned up code of previous PRs that all went into the `0-19-staging` branch and represents all the changes needed to get custom channels into `lnd`.

# Reviewers please read

All commits have been pre-reviewed in previous PRs by at least one person.
We are looking for a second round review, mostly with focus on safety and stability when running with non-custom channels.

The idea is that reviewers get assigned to a certain range of commits within this PR. We decided to _not_ do multiple PRs to keep rebase effort minimal (and so we have all context in a single PR).

The following parts exist (see commit message prefixes marking the start of a part):
 - `-PART1--`: General tapscript leaf awareness and aux leaf store
    - Assigned reviewers: @ProofOfKeags 
    - :green_circle:  **Status**: DONE.
    - EXTRACTED from this PR into: https://github.com/lightningnetwork/lnd/pull/9030
 - `-PART2--`: Custom records on HTLCs, SCID alias management, TLV traffic shaper
    - Assigned reviewers: @bitromortac @ProofOfKeags 
    - :green_circle:  **Status**: DONE.
    - EXTRACTED from this PR into: https://github.com/lightningnetwork/lnd/pull/9049
 - `-PART3--`: Aux funding and RPC/CLI custom data integration
    - Assigned reviewers: @Crypt-iQ @Roasbeef 
    - :green_circle:  **Status**: DONE.
    - EXTRACTED from this PR into: https://github.com/lightningnetwork/lnd/pull/9072
 - `-PART4--`: Invoice HTLC modifier and aux channel closer
    - Assigned reviewers: @ellemouton @bhandras 
    - :green_circle:  **Status**: DONE.
    - EXTRACTED from this PR into: https://github.com/lightningnetwork/lnd/pull/9095
 - `-PART5--`: Force close sweeping
    - Assigned reviewers: @yyforyongyu @Crypt-iQ 
    - 🛠️ **Status**: Ready for full review, review in progress...


# Included changes from recent PRs to the `0-19-staging` branch

The following PRs **HAVE** been integrated either into a previous part or this last, 5th part:
 - https://github.com/lightningnetwork/lnd/pull/9092
 - https://github.com/lightningnetwork/lnd/pull/9120
 - https://github.com/lightningnetwork/lnd/pull/9032
 - https://github.com/lightningnetwork/lnd/pull/9040
 - https://github.com/lightningnetwork/lnd/pull/8947
 - https://github.com/lightningnetwork/lnd/pull/8926

## Not included PRs to the `0-19-staging` branch

The following PRs have **NOT** yet been integrated into this PR (as they're currently in-flight):
 - https://github.com/lightningnetwork/lnd/pull/9074 (only partially included, latest fixes need to be re-applied)
 - https://github.com/lightningnetwork/lnd/pull/9110

Roasbeef commented:
> The following parts exist (see commit message prefixes marking the start of a part):

This prefix seems to have been dropped in the latest rebase?

guggero commented:
> This prefix seems to have been dropped in the latest rebase?

Oops, I haven't actually pushed it yet... It's now up to date and also rebased on top of the first refactor work in the channel state machine.

guggero commented:
I addressed some of the TODOs and fixes the unit and integration test failures. Also added release notes that cover all new functionality of this PR.

I added a "Status" section to each part of this PR in the main PR body, will update that whenever status changes.
That means part 1, 3 and 5 are fully ready for review, @ProofOfKeags, @Crypt-iQ, @yyforyongyu.

yyforyongyu reviewed (commented):
Did a quick review on part 5, commits look pretty good. Will do a deep dive, meanwhile let me know if you need help patching unit tests and itest, especially for packages `sweep` and `contractcourt`.

ProofOfKeags commented on contractcourt/chain_watcher.go:
```diff
@@ -446,11 +461,16 @@ func (c *chainWatcher) handleUnknownLocalState(
 	// Next, we'll derive our script that includes the revocation base for
 	// the remote party allowing them to claim this output before the CSV
 	// delay if we breach.
+	localAuxLeaf := fn.MapOption(
+		func(l lnwallet.CommitAuxLeaves) input.AuxTapLeaf {
+			return l.LocalAuxLeaf
+		},
+	)(auxLeaves)
```
map followed by flatten is just [`ChainOption`](https://github.com/lightningnetwork/lnd/blob/master/fn/option.go#L136-L148)

ProofOfKeags commented on lnwallet/channel.go:
```diff
@@ -692,10 +695,21 @@ func (lc *LightningChannel) extractPayDescs(feeRate chainfee.SatPerKWeight,
 
 		htlc := htlc
 
+		auxLeaf := fn.MapOption(
+			func(l CommitAuxLeaves) input.AuxTapLeaf {
+				leaves := l.OutgoingHtlcLeaves
+				if htlc.Incoming {
+					leaves = l.IncomingHtlcLeaves
+				}
+
+				return leaves[htlc.HtlcIndex].AuxTapLeaf
+			},
+		)(auxLeaves)
```
Same here.

ProofOfKeags commented on lnwallet/channel.go:
```diff
@@ -692,10 +695,21 @@ func (lc *LightningChannel) extractPayDescs(feeRate chainfee.SatPerKWeight,
 
+		auxLeaf := fn.MapOption(
+			func(l CommitAuxLeaves) input.AuxTapLeaf {
+				leaves := l.OutgoingHtlcLeaves
+				if htlc.Incoming {
+					leaves = l.IncomingHtlcLeaves
+				}
```
This pattern gives rise to the idea that you should pack these things into [`Duals`](https://github.com/lightningnetwork/lnd/pull/8996) and then use a `ChannelParty` to index into the Dual. This will reduce these 4 lines to a one-liner that is simpler and less error prone due to single-assignment.

guggero replied:
I agree to use `Dual` for things like local and remote balance or local and remote keys. But incoming/outgoing IMO doesn't directly map to local/remote. AFAIK incoming/outgoing is always static as seen from the node processing the commit (but independent of **whose** commit is being processed).

ProofOfKeags replied:
if something is incoming it means that the origin is the remote and that the destination is local. So you could frame as "htlcLeavesOrigins" or "htlcLeavesDestinations". The mapping is definitely sound, it's just a matter of whether you want to use the tools.

ProofOfKeags commented on lnwallet/channel.go:
```diff
@@ -742,13 +756,27 @@ func (lc *LightningChannel) diskCommitToMemCommit(
 
+	auxLeaves, err := AuxLeavesFromCommit(
+		lc.channelState, *diskCommit, lc.leafStore,
+		func() CommitmentKeyRing {
+			if whoseCommit.IsLocal() {
+				return *localCommitKeys
+			}
+
+			return *remoteCommitKeys
```
If you pack local and remote commit keys into a dual, this just becomes `commitKeys.GetParty(whoseCommit)`

guggero replied:
Added a refactor commit.

ProofOfKeags reviewed (changes requested):
Ok so the biggest thing I want to challenge is the idea that we should use an Option for the tapscript root. IIRC we always have a tapscript root, irrespective of the presence of any aux leaves. We may not have any aux leaves and so I think it makes sense for that to be an option but we can probably save ourselves the effort and just compute the BIP86 root during initialization rather than computing it at the end and having to thread through optional arguments everywhere. The more cases we can eliminate earlier the better, I think. I'm open to some pushback here, though.

Other than that I have some SWE notes intended to simplify some of the interface/implementation code here.

NOTE: my review ONLY covers **PART 1**

Roasbeef commented:
I pulled in the msg router specific comments from this PR into the original PR adding it: https://github.com/lightningnetwork/lnd/pull/8520

Ideally we can get this one in quickly, then rebase this over it.

Roasbeef commented on sweep/tx_input_set.go:
```diff
@@ -220,6 +220,26 @@ func (b *BudgetInputSet) NeedWalletInput() bool {
+	if extraRequiredTxOut {
+		// TODO(roasbeef): aux sweeper ext to ask for extra output
+		// params and value?
+		budgetNeeded += 1_000
```
In `tapd` rn, we use a value of 1k sats for the top level anchor transaction: https://github.com/lightninglabs/taproot-assets/blob/8fcd27c156b9628304a7c098ca61a22910d9a869/tapsend/send.go#L35-L40

So if we're adding a sweep output, we need to commit an extra 1k sats.

Roasbeef commented on input/input.go:
```diff
@@ -182,10 +199,11 @@ func MakeBaseInput(outpoint *wire.OutPoint, witnessType WitnessType,
 func NewBaseInput(outpoint *wire.OutPoint, witnessType WitnessType,
-	signDescriptor *SignDescriptor, heightHint uint32) *BaseInput {
+	signDescriptor *SignDescriptor, heightHint uint32,
+	opts ...InputOpt) *BaseInput {
```
The idea is that this is a more general pattern to pave the way for adding other params in.

Roasbeef commented on lnwallet/interface.go:
```diff
@@ -594,6 +596,67 @@ type MessageSigner interface {
+	// If it's not a taproot address, we don't require to know the internal
+	// key in the first place. So we don't return an error here, but also no
+	// internal key.
+	if !txscript.IsPayToTaproot(deliveryScript) {
```
IIRC we did initially, but then that needed special handling at the caller side. In this case, we return `None`, as there just isn't an internal key for non taproot addrs.

guggero replied:
Yeah, we call this user defined delivery addresses (e.g. upfront shutdown address). So we can't error out, as that would break the unit and integration tests.
I think we'll eventually need to find a workaround if we want to support upfront shutdown addresses for Taproot Asset Channels...

Roasbeef commented on contractcourt/briefcase.go:
```diff
@@ -1554,6 +1556,12 @@ func encodeTaprootAuxData(w io.Writer, c *ContractResolutions) error {
 		//nolint:lll
 		tapCase.CtrlBlocks.Val.CommitSweepCtrlBlock = commitSignDesc.ControlBlock
+
+		c.CommitResolution.ResolutionBlob.WhenSome(func(b []byte) {
```
If there's nothing there, then we just don't do anything, as there's nothing to encode/write.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -2188,6 +2216,35 @@ func NewBreachRetribution(chanState *channeldb.OpenChannel, stateNum uint64,
+		resolveBlob := fn.MapOptionZ(
+			auxResolver,
+			func(a AuxContractResolver) fn.Result[tlv.Blob] {
+				//nolint:lll
```
You mean this `lll` comment, or something else?

Roasbeef commented:
@ProofOfKeags 

> Ok so the biggest thing I want to challenge is the idea that we should use an Option for the tapscript root. IIRC we always have a tapscript root, irrespective of the presence of any aux leaves

We don't always hve a tapscript root. When we do musig2 today for taproot channels, we don't have that extra tweak, so it's effectively BIP 86. This ensures that the only way the output can be spent is via the musig2 multi-sig with no script path.

Roasbeef commented on lnwallet/chanfunding/interface.go:
```diff
@@ -119,6 +121,11 @@ type Request struct {
 	// funding output.
 	Musig2 bool
+
+	// TapscriptRoot is the root of the tapscript tree that will be used to
+	// create the funding output. This field will only be utilized if the
+	// Musig2 flag above is set to true.
+	TapscriptRoot fn.Option[chainhash.Hash]
```
So this is for the funding output. If you're doing vanilla taproot chans, there's no tapscript root.

ProofOfKeags replied:
The tapscript root is the BIP86 root though isn't it?

guggero replied:
See https://github.com/lightningnetwork/lnd/pull/8960#issuecomment-2296096887.

Roasbeef commented on lnwallet/chanfunding/canned_assembler.go:
```diff
@@ -56,6 +58,14 @@ type ShimIntent struct {
+	// create the funding output. This field will only be utilized if the
+	// MuSig2 flag above is set to true.
+	//
+	// TODO(roasbeef): fold above into new chan type? sum type like thing,
+	// includes the tapscript root, etc
```
In this case, the type (eg: `musig2`) isn't sufficient. As we can have that be true, but expect no tapscript root. We need the tapscript root for the aux chans when it's present.

Roasbeef commented on lnwallet/wallet.go:
```diff
@@ -2459,6 +2459,12 @@ func initStateHints(commit1, commit2 *wire.MsgTx,
 	return nil
 }
 
+// TapscriptRootToOpt is a helper function that converts a tapscript root into
+// the functional option we can use to pass into GenTaprootFundingScript.
+func TapscriptRootToOpt(root chainhash.Hash) []input.FundingScriptOpt {
+	return []input.FundingScriptOpt{input.WithTapscriptRoot(root)}
+}
```
Yep you're right, not sure why I added the slice initially.

guggero replied:
I think this is to work nicely with `MapOptionZ` and the spread operator.

Example:
```
		fundingOpts := fn.MapOptionZ(
			channel.TapscriptRoot, lnwallet.TapscriptRootToOpt,
		)
		pkScript, _, err := input.GenTaprootFundingScript(
			localKey, remoteKey, int64(channel.Capacity),
			fundingOpts...,
		)
```

If `channel.TapscriptRoot` is `None`, then `fundingOpts` is `Zero` (which for a slice is `nil`, which is "spreadable").
If it's some, then the result is a slice with size 0.

ProofOfKeags replied:
Yeah it just makes the components less composable which just makes the implementation more cumbersome in the long run imo. I'm not attached to this. Feel free to resolve as "won't fix" if you disagree.

Roasbeef commented on lnwallet/test_utils.go:
```diff
@@ -348,6 +349,21 @@ func CreateTestChannels(t *testing.T, chanType channeldb.ChannelType,
+
+		someRoot := fn.Some(tapscriptRoot)
+
+		aliceChannelState.TapscriptRoot = someRoot
+		bobChannelState.TapscriptRoot = someRoot
```
You mean why isn't this handled by the traditional constructors for the channel state?

Elsewhere this is set when we're making the `ChannelReservation` for the first time, but invoking those functions are out of the scope of the unit test.

Roasbeef commented on input/script_utils.go:
```diff
@@ -2097,6 +2123,12 @@ type CommitScriptTree struct {
+	// AuxLeaf is an auxiliary leaf that can be used to extend the base
+	// commitment script tree with new spend paths, or just as extra
+	// commitment space. When present, this leaf will always be in the
+	// left-most or right-most area of the tapscript tree.
```
This is a vague allusion to the tapscript leaf uniqueness rules that `tapd` uses. We want to ensure only a single commitment exists in the tree.

guggero replied:
IIRC in practice we ended up always adding it as the right-most leaf. Changed the comment to reflect that and be less confusing.

Roasbeef commented on channeldb/channel.go:
```diff
@@ -252,6 +252,10 @@ type chanAuxData struct {
 	// output commits to.
 	tapscriptRoot tlv.OptionalRecordT[tlv.TlvType6, [32]byte]
+
+	// customBlob is an optional TLV encoded blob of data representing
+	// custom channel funding information.
+	customBlob tlv.OptionalRecordT[tlv.TlvType7, tlv.Blob]
```
The pattern in this PR is generally to pass down this opaque blob all the way until something external to `lnd` (in this case `tapd`) knows how to map it into some concrete type. 

In this case, `tapd` passes us some serialized state, and we store it. This way it doesn't need to roll new datastores for the channels, and we can piggy back onto the existing state flow for chans.

ProofOfKeags replied:
Yeah that makes sense.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -2463,18 +2463,29 @@ func HtlcIsDust(chanType channeldb.ChannelType,
-	ourUpdates   []*PaymentDescriptor
-	theirUpdates []*PaymentDescriptor
-	feePerKw     chainfee.SatPerKWeight
+type HtlcView struct {
+	// NextHeight is the height of the commitment transaction that will be
+	// created using this view.
+	NextHeight uint64
+
+	// OurUpdates are our outgoing HTLCs.
+	OurUpdates []*PaymentDescriptor
+
+	// TheirUpdates are their incoming HTLCs.
+	TheirUpdates []*PaymentDescriptor
+
+	// FeePerKw is the fee rate in sat/kw of the commitment transaction.
+	FeePerKw chainfee.SatPerKWeight
```
Yeah...no way to avoid other than prioritization. Sooner we get this in, sooner we don't need to continually rebase a PR of this size.

guggero replied:
:saluting_face:

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -6290,12 +6323,14 @@ func NewUnilateralCloseSummary(chanState *channeldb.OpenChannel, signer input.Si
 	commitTxBroadcast := commitSpend.SpendingTx
 
+	// TODO(roasbeef): fetch aux leave
+
```
I think there were some commits dropped here @guggero? 

This is what the file looks like in the OG branch: https://github.com/lightningnetwork/lnd/blob/1b353b0bfd58a6dfeafc2878111404048826b41d/lnwallet/channel.go#L7157-L7162

Roasbeef replied:
Or actually this is on a dated part of the diff, it's correct here: https://github.com/lightningnetwork/lnd/blob/626d901449e18ee0f96d2977677581ae4cd6a5df/lnwallet/channel.go#L6659-L6664

guggero replied:
Updated the comment to make it more clear this is going to be fixed in a later commit of this PR.

@Roasbeef commits should all be there, I verified this against the `litd` integration tests at some point.

Roasbeef commented on lnwallet/commitment.go:
```diff
@@ -610,30 +612,108 @@ func CommitScriptAnchors(chanType channeldb.ChannelType,
 	return localAnchor, remoteAnchor, nil
 }
 
+// CommitSortFunc is a function type alias for a function that sorts the
+// commitment transaction outputs. The second parameter is a list of CLTV
+// timeouts that must correspond to the number of transaction outputs, with the
+// value of 0 for non-HTLC outputs.
+type CommitSortFunc func(*wire.MsgTx, []uint32, []input.HtlcIndex) error
```
Do you mean not using mutation to do the sort? FWIW that's how things are with the official sort function: https://pkg.go.dev/sort#Sort

ProofOfKeags replied:
I'm talking about sorting the htlcs according to the sort rules before rendering them to outputs, rather than rendering to outputs and using parallel vectors to do an in-place sort.

ProofOfKeags replied:
Maybe I'll just clean this up myself at a later date though.

guggero replied:
Yeah, I think this would explode the scope of this PR. Though I agree this could probably be done in a better way.

Roasbeef commented on channeldb/revocation_log.go:
```diff
@@ -72,116 +140,60 @@ var (
 type HTLCEntry struct {
 	// RHash is the payment hash of the HTLC.
-	RHash [32]byte
+	RHash tlv.RecordT[tlv.TlvType0, SparsePayHash]
```
It's always set, note that we're not using `tlv.OptionalRecordT` here, we're using `tlv.RecordT`.

ProofOfKeags replied:
Right but when is it ever going to be the sparse prunable value?

Roasbeef commented on chainreg/chainregistry.go:
```diff
@@ -63,6 +64,10 @@ type Config struct {
 	// state.
 	ChanStateDB *channeldb.ChannelStateDB
 
+	// AuxLeafStore is an optional store that can be used to store auxiliary
+	// leaves for certain custom channel types.
+	AuxLeafStore fn.Option[lnwallet.AuxLeafStore]
```
I think this is out of place, or just unnecessary. We instead made an `AuxComponents` struct in `config_builder.go`.

guggero replied:
We need it in `BuildChainControl()` to pass into the `lnwallet.Config{}` struct which is needed for `lnwallet.NewLightningWallet()`.

ProofOfKeags replied:
Weird that the chain control depends on lnwallet. But that's out of scope for this PR as it's deeper in the arch.

ProofOfKeags commented:
> FWIW, the BIP 86 root is basically just `nil`

Why don't we just compute the BIP86 value and drop it into that structure so that we can unify the types without needing the Option? I'm trying to push for symmetry/uniformity and I think there is natural symmetry here since there is always a value for this. I just think in the nil case you're computing it later?

guggero commented:
> Why don't we just compute the BIP86 value and drop it into that structure so that we can unify the types without needing the Option? I'm trying to push for symmetry/uniformity and I think there is natural symmetry here since there is always a value for this. I just think in the nil case you're computing it later?

The BIP86 script root is an _empty_ byte slice: https://github.com/btcsuite/btcd/blob/master/txscript/taproot.go#L290
We can't put that into a `chainhash.Hash`, unless we use the all-zero hash. But then we need another distinction somewhere, which IMO is more dangerous.

guggero commented on go.mod:
```diff
@@ -41,7 +41,7 @@ require (
 	github.com/lightningnetwork/lnd/sqldb v1.0.3
 	github.com/lightningnetwork/lnd/ticker v1.1.1
-	github.com/lightningnetwork/lnd/tlv v1.2.3
+	github.com/lightningnetwork/lnd/tlv v1.2.6
```
You mean pull it out into a separate PR? I guess I could extract some independent commits into its own PR and get that merged sooner. But the initial plan was to get everything in as a whole...

ProofOfKeags replied:
I'm fine with leaving it where it is. I'm mostly saying this because I know I've had trouble getting large PRs in and things go faster when smaller chunks are made. If you wanna leave it where it is go for it.

guggero replied:
Ended up extracting it into its own PR.

guggero commented:
Rebased after https://github.com/lightningnetwork/lnd/pull/8520 and addressed some first comments.

ProofOfKeags commented on routing/router.go:
```diff
@@ -290,6 +292,10 @@ type Config struct {
 	// TODO(yy): remove it once the root cause of stuck payments is found.
 	ClosedSCIDs map[lnwire.ShortChannelID]struct{}
+
+	// TrafficShaper is an optional traffic shaper that can be used to
+	// control the outgoing channel of a payment.
+	TrafficShaper fn.Option[TlvTrafficShaper]
```
Since the traffic shaper is itself an interface, you may have better success giving noop semantics to the nil pointer and then not using an Option here. The ability to do this hangs on whether you can give a satisfactory interpretation of the request if the shaper is nil that doesn't break the promised semantics of the interface.

Roasbeef replied:
Hmm, I don't think we can enforce that just at the _interface_ level, instead all the implementations would need to handle the nil pointer receive case to return _some_ value. But with the option format, we detect that by just having a `None` option which is then mapped to `None`.

ProofOfKeags commented on routing/payment_lifecycle.go:
```diff
@@ -686,6 +688,38 @@ func (p *paymentLifecycle) sendAttempt(
+				"records %v and amount %d msat for HTLC",
+				spew.Sdump(customRecords), newAmt)
+
+			htlcAdd.CustomRecords = customRecords
+			htlcAdd.Amount = newAmt
```
once again I *strongly* discourage mutating outer scope values in these closures. If you insist on doing so, don't use the Option type, since the general expectation about the usage of MapOptionZ is to not have side effects.

ProofOfKeags reviewed (changes requested):
Just completed a first pass of the PART2 Review. My main concerns are related to fixing the alias map desynchronization using a better (more authoritative/canonical) means of managing the aliases throughout LND. Other than that I want to put in a warning here that there will be conflicts with #8981. The rest of my comments are more SWE/impl related and can be deprioritized until we resolve some of the larger design questions.

yyforyongyu commented on routing/unified_edges.go:
```diff
@@ -250,7 +250,7 @@ func (u *edgeUnifier) getEdgeLocal(netAmtReceived lnwire.MilliSatoshi,
 
 		// Check valid amount range for the channel.
-		if !edge.amtInRange(amt) {
+		if htlcBlob.IsNone() && !edge.amtInRange(amt) {
```
won't this HTLC be rejected by our downstream peer?

guggero replied:
If the downstream node isn't a node with custom channel support, then yes. But we'll only send out such HTLCs to custom channels, so the peers should understand the difference in value between the actual on-chain amount and what's potentially being transported in the HTLC. Added a comment to make more clear though.

yyforyongyu commented on routing/pathfind.go:
```diff
@@ -1447,3 +1473,14 @@ func lastHopPayloadSize(r *RestrictParams, finalHtlcExpiry int32,
+func overflowSafeAdd(x, y lnwire.MilliSatoshi) lnwire.MilliSatoshi {
+	if y > math.MaxUint64-x {
+		// Overflow would occur, return maximum uint64 value.
+		return math.MaxUint64
```
think we should panic here

guggero replied:
Returned a boolean instead to allow the caller to handle the situation.

yyforyongyu commented on routing/payment_lifecycle.go:
```diff
@@ -686,6 +688,38 @@ func (p *paymentLifecycle) sendAttempt(
 
+	// If a hook exists that may affect our outgoing message, we call it now
+	// and apply its side effects to the UpdateAddHTLC message.
+	err := fn.MapOptionZ(
```
Can we create a method instead of bloating this method

yyforyongyu commented on routing/bandwidth.go:
```diff
@@ -18,7 +22,40 @@ type bandwidthHints interface {
+// TlvTrafficShaper is an interface that allows the sender to determine if a
+// payment should be carried by a channel based on the TLV records that may be
+// present in the `update_add_htlc` message or the channel commitment itself.
+type TlvTrafficShaper interface {
```
I think this belongs to the `link`?

guggero replied:
No, I don't think so. This clearly affects decisions made by the router. We can discuss the name if you have any ideas. But the tasks for this component are: Determine and choose what outbound channels can be used for a specific outgoing payment, then produce custom records for the first hop of such a payment.

yyforyongyu commented on .golangci.yml:
```diff
@@ -259,8 +259,8 @@ issues:
 
-    # Allow fmt.Printf() in lncli.
-    - path: cmd/lncli/*
+    # Allow fmt.Printf() in commands.
```
think this commit can be easily reviewed and merged in its own PR.

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -3035,14 +3029,8 @@ func processRemoveEntry(htlc *PaymentDescriptor, ourBalance,
-	}
-
 	// Ignore any removal entries which have already been processed.
+	removeHeight := RemoveHeight(htlc, whoseCommitChain)
```
`RemoveHeight` -> `RemoteHeight`

guggero replied:
No, this is actually the height an HTLC was removed at.

yyforyongyu commented on input/script_desc.go:
```diff
@@ -73,6 +73,9 @@ type TapscriptDescriptor interface {
 	TapScriptTree() *txscript.IndexedTapScriptTree
+
+	// Tree returns the underlying ScriptTree.
+	Tree() ScriptTree
```
Q: is `TapScriptTree` a subset of `Tree`?

guggero replied:
No, the `*txscript.IndexedTapScriptTree` is the parsed leaves in the structure that can then build the root hash. The `ScriptTree` is something every `TapscriptDescriptor` implementation of the `ScriptDescriptor` has embedded, but before this change there was no easy way to access it.

Pulled this commit out of the PR, since it's kind of standalone.

Roasbeef replied:
@guggero the `Tree()` method is used in `tapchannel.LeavesFromTapscriptScriptTree`.

yyforyongyu commented on rpcserver_test.go:
```diff
@@ -1,14 +1,79 @@
 }
+
+// mockDataParser is a mock implementation of the AuxDataParser interface.
+type mockDataParser struct {
```
use `mock.Mock` instead?

guggero replied:
Same answer as above: Happy to help do that outside the context of this PR.

guggero commented:
I extracted some commits into https://github.com/lightningnetwork/lnd/pull/9025 as requested.

All TODOs except for a single one (will just be an additional itest) are addressed, so all parts are now officially ready for full review.

@saubyk I'll update the PR body with the status regularly, please always look at that first.

guggero commented on lnwire/extra_bytes_test.go:
```diff
@@ -143,3 +143,66 @@ func TestExtraOpaqueDataPackUnpackRecords(t *testing.T) {
+// ExtraOpaqueData instance, and then extract them back out. Crucially, we'll
+// ensure that records can be packed in any order, and we'll ensure that the
+// unpacked records are valid.
+func TestPackRecords(t *testing.T) {
```
Yes, I'll keep that in mind for future unit tests.

guggero commented on lnwire/update_add_htlc_test.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+// generateTestCases generates a set of UpdateAddHTLC message test cases.
+func generateTestCases(t *testing.T) []testCase {
```
Probably due to the commit author not being familiar with either. Would count myself into that camp as well. Will invest some time into learning one of them (I assume `rapid` would be the more obvious choice since it probably is more feature rich?).

ProofOfKeags replied:
Yeah the consensus is that new tests should move towards rapid. We have plenty of examples using quick right now to get an intuition for test design. However, quick has been abandoned and rapid has better support. CC @Roasbeef

guggero commented on htlcswitch/interceptable_switch.go:
```diff
@@ -129,6 +129,10 @@ type FwdResolution struct {
 	// FwdActionSettle.
 	Preimage lntypes.Preimage
 
+	// IncomingAmountMsat is the amount that is to be used for validating if
+	// Action is FwdActionResumeModified.
+	IncomingAmountMsat fn.Option[lnwire.MilliSatoshi]
```
Yes, actually they are used in completely different scenarios. You'd likely only ever modify one. For an `asset -> BTC` forward you'd update the `IncomingAmountMsat` to tell the validation logic it's fine to accept the incoming BTC amount in the HTLC since the actual "worth" of the HTLC is `IncomingAmountMsat`.

For a `BTC -> asset` forward, you set `OutgoingAmountMsat` to tell the forwarding logic to only put `OutgoingAmountMsat` in the on-chain HTLC, even though the actual "worth" of the HTLC as a whole might be way more.

guggero commented on lnwallet/commitment.go:
```diff
@@ -807,7 +976,11 @@ func (cb *CommitmentBuilder) createUnsignedCommitmentTx(ourBalance,
+
+		// We want to add the CLTV and HTLC index to their respective
+		// slices, even if we already pre-allocated them.
+		cltvs = append(cltvs, htlc.Timeout)               //nolint
```
Unfortunately we need it. Otherwise it complains about `make([]uint32, len(commitTx.TxOut))` (missing `0` secondary arg) and `append()` being used together (which often is wrong but in this specific context is what we want to do).
Comment above was supposed to explain this, but can be more clear if that's imprecise.

yyforyongyu commented on invoices/settlement_interceptor.go:
```diff
@@ -0,0 +1,224 @@
+// invoice is submitted to this service. This session allows the caller to block
+// until the invoice is processed.
+type InterceptSession struct {
+	HtlcModifyRequest
```
I find it structurally weird to embed a session with a request, can we do `Req HtlcModifyRequest` instead?

yyforyongyu commented on invoices/settlement_interceptor.go:
```diff
@@ -0,0 +1,224 @@
+		ClientErrChannel:      make(chan error, 1),
+		Quit:                  make(chan struct{}, 1),
+	}
+	s.activeSessions[sessionKey] = session
```
Is it safe to overwrite? Could there exist a session under the same key?

yyforyongyu commented on invoices/settlement_interceptor.go:
```diff
@@ -0,0 +1,224 @@
+// Stop stops the service.
+func (s *HtlcModificationInterceptor) Stop() error {
+	// If the service is stopping, we will quit all active sessions.
+	return s.QuitActiveSessions()
```
missing `s.wg.Wait()`

yyforyongyu commented on invoices/settlement_interceptor.go:
```diff
@@ -0,0 +1,224 @@
+var _ HtlcModifier = (*HtlcModificationInterceptor)(nil)
+
+// MockHtlcModifier is a mock implementation of the HtlcModifier interface.
+type MockHtlcModifier struct {
```
this should be put in a `_test.go` file or `mock.go`

yyforyongyu commented on invoices/invoiceregistry.go:
```diff
@@ -1027,6 +1032,53 @@ func (i *InvoiceRegistry) notifyExitHopHtlcLocked(
+		// with a settlement resolution.
+		var interceptErr error
+		interceptSession.WhenSome(func(session InterceptSession) {
+			log.Debug("Waiting for client response from invoice " +
```
it may be helpful to log the session key in case there are multiple sessions.

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -7721,10 +7721,21 @@ func NewLocalForceCloseSummary(chanState *channeldb.OpenChannel,
 	}, nil
 }
 
+// CloseOutput is wraps a normal tx out with additional metadata that indicates
```
```suggestion
// CloseOutput wraps a normal tx out with additional metadata that indicates
```

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -7721,10 +7721,21 @@ func NewLocalForceCloseSummary(chanState *channeldb.OpenChannel,
 }
 
+// CloseOutput is wraps a normal tx out with additional metadata that indicates
+// if the output to the initiator of the channel or not.
```
```suggestion
// if the output belongs to the initiator of the channel or not.
```

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -8628,17 +8673,101 @@ func CreateCooperativeCloseTx(fundingTxIn wire.TxIn,
+		switch {
+		// For additional local outputs, add the output, then deduct
+		// the balance from our local balance.
+		case extraTxOut.IsLocal:
```
is it possible that we have a local extra output but the `localOutputIdx` is none?

guggero replied:
Yes, I think it is. Added this comment to make it more clear:
```
			// The extraCloseOutputs in the options just indicate if
			// an extra output should be added in general. But we
			// only add one if we actually _need_ one, based on the
			// balance. If we don't have enough local balance to
			// cover the extra output, then localOutputIdx is None.
```

yyforyongyu commented on lnwallet/channel_test.go:
```diff
@@ -11120,3 +11121,315 @@ func TestBlindingPointPersistence(t *testing.T) {
+
+// TestCreateCooperativeCloseTx tests that the cooperative close transaction is
+// properly created based on the standard and also optional parameters.
+func TestCreateCooperativeCloseTx(t *testing.T) {
```
nice👍

guggero commented on channeldb/revocation_log.go:
```diff
@@ -164,22 +164,32 @@ type HTLCEntry struct {
+
+	// CustomBlob is an optional blob that can be used to store information
+	// specific to revocation handling for a custom channel type.
+	CustomBlob tlv.OptionalRecordT[tlv.TlvType5, tlv.Blob]
```
Yeah, this won't be super light on the hard disk. This blob will contain a proof suffix, which can easily be a couple of hundred bytes. Will need to do some optimizations at some point. But finish all the functionality first I guess.

guggero commented on lnwire/extra_bytes.go:
```diff
@@ -105,6 +140,45 @@ func (e *ExtraOpaqueData) ExtractRecords(recordProducers ...tlv.RecordProducer)
+	var recordProducers []tlv.RecordProducer
+
+	// If the instance is nil or empty, return an empty slice.
+	if e == nil || len(*e) == 0 {
```
No, Go also does lazy evaluation, if the first part of an `OR` operation results in `true`, the second part isn't evaluated. https://stackoverflow.com/questions/53722288/short-circuit-evaluation-in-go

yyforyongyu replied:
what i meant was, since `e` is nil, then calling `e.RecordProducers` would panic, hence the `e` can't be nil?

guggero replied:
No, that's the cool thing in Go. You can call a method on a `nil` object. And as long as you don't access any members of the object (and properly handle the `nil` case as we do here), there is no nil pointer panic.

yyforyongyu replied:
oh yeah right, remember having a similar conversation with you before...deja vu

guggero commented on lnwire/update_add_htlc.go:
```diff
@@ -72,6 +73,11 @@ type UpdateAddHTLC struct {
+	// CustomRecords maps TLV types to byte slices, storing arbitrary data
+	// intended for inclusion in the ExtraData field of the UpdateAddHTLC
+	// message.
+	CustomRecords CustomRecords
```
The `ExtraData` is only used for serialization now. It will carry `BlindingPoint` + `CustomRecords` when put on the wire or into the DB.

Roasbeef commented on htlcswitch/interceptable_switch.go:
```diff
@@ -123,6 +129,14 @@ type FwdResolution struct {
 	// FwdActionSettle.
 	Preimage lntypes.Preimage
 
+	// OutgoingAmountMsat is the amount that is to be used for forwarding if
+	// Action is FwdActionResumeModified.
+	OutgoingAmountMsat fn.Option[lnwire.MilliSatoshi]
+
+	// CustomRecords is the custom records that are to be used for
+	// forwarding if Action is FwdActionResumeModified.
+	CustomRecords fn.Option[record.CustomSet]
```
You can have both in the general set: this can be used by "LSPs" to implement JIT channels where they skim off some of the incoming HTLCs as fees.

Roasbeef commented on itest/lnd_forward_interceptor_test.go:
```diff
@@ -344,6 +345,143 @@ func testForwardInterceptorBasic(ht *lntest.HarnessTest) {
+		return strings.Contains(logEntry, targetLogPrefixStr) &&
+			strings.Contains(logEntry, targetCustomRecordsStr) &&
+			strings.Contains(logEntry, targetOutgoingAmountMsatStr)
+	}
```
This was a temp itest before we go everything full working e2e. So I think we may just be able to remove it at this point, given we have extra coverage elsewhere.

Roasbeef commented on htlcswitch/switch.go:
```diff
@@ -2090,6 +2090,26 @@ func (s *Switch) addLiveLink(link ChannelLink) {
+	s.updateLinkAliases(link)
+}
+
+// UpdateLinkAliases is the externally exposed wrapper for updating link
+// aliases. It acquires the indexMtx and calls the internal method.
+func (s *Switch) UpdateLinkAliases(link ChannelLink) {
+	s.indexMtx.Lock()
+	defer s.indexMtx.Unlock()
+
+	s.updateLinkAliases(link)
+}
+
+// updateLinkAliases updates the aliases for a given link. This will cause the
+// htlcswitch to consult the alias manager on the up to date values of its
+// alias maps.
+//
+// NOTE: this MUST be called with the indexMtx held.
+func (s *Switch) updateLinkAliases(link ChannelLink) {
+	linkScid := link.ShortChanID()
+
```
Yeah this was a bigger change we considered at the time, it partially was spun off into this issue: https://github.com/lightningnetwork/lnd/issues/8969. 

Right now the switch maintains a series of maps to mirror the state of the alias manager. Instead the alias manager should fully encapsulate all that state, ensuring that we only have a single source of truth. 

I don't think we're able to commit to making this architectural change in the scope of this PR, given the existing logic here, and the downstream items that we still need to address (cherry pick on top of 0.18.3 after merging into master, merging in changes into the two branches, etc). 

I think @GeorgeTsagk may have a side branch where he set out to bite the bullet and make the larger change.

ProofOfKeags replied:
Ok. I'll defer regarding scope but as long as we are in agreement that this is Bad™ then I can lay this to rest. That said it does make it hard to assess the correctness of the change since we cannot reason locally about the behavior and I don't have a deep knowledge of how this state synchronizes using the disk.

Roasbeef commented on routing/bandwidth.go:
```diff
@@ -89,30 +130,104 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
+		log.Debugf("ShortChannelID=%v: external traffic shaper "+
+			"reported available bandwidth: %v", cid, auxBandwidth)
+
+		auxBandwidthDetermined = true
```
In this case, it's the inner scope of the closure from the option, I think the only way would be to return a full struct here.

ProofOfKeags replied:
Yes that's the correct way to handle it. If you want to return an anonymous pair we do have `fn.T2`. But if you have more values I'd recommend a struct.

guggero replied:
Yeah, makes sense. Completely refactored this to use a `fn.Result()`. I start to see the elegance of that type...

ProofOfKeags replied:
It's one of those things where it gets more useful the more you use it. All these `fn` primitives compose together really well. Its benefits compound.

Roasbeef commented on contractcourt/chain_watcher.go:
```diff
@@ -423,15 +432,30 @@ func (c *chainWatcher) handleUnknownLocalState(
 		&c.cfg.chanState.LocalChanCfg, &c.cfg.chanState.RemoteChanCfg,
 	)
 
+	auxLeaves, err := lnwallet.AuxLeavesFromCommit(
```
This is done indirectly in `AuxLeavesFromCommit` right now, but yeah could make sense to put it up to this level. Would also let us eliminate usage of `UnsafeSome()` there.

guggero replied:
Refactored this.

Roasbeef commented on channeldb/revocation_log.go:
```diff
@@ -155,19 +155,17 @@ type HTLCEntry struct {
 	CustomBlob tlv.OptionalRecordT[tlv.TlvType5, tlv.Blob]
+
+	// HtlcIndex is the index of the HTLC in the channel.
+	HtlcIndex tlv.RecordT[tlv.TlvType6, uint16]
```
Hmm, trying to remember why I added this....not sure it's actually used as is.

Roasbeef replied:
Nvm, it's used in `createHtlcRetribution` in this diff. To extract the proper aux leaf for a given HTLC we want to resolve. 

Perhaps we can make it `OptionalRecordT` here instead. FWIW, if the aux chan stuff isn't active, it isn't used at all.

guggero replied:
Hmm, yeah, old records won't have this value, so we'll decode it as 0. Which might not always be the desired outcome and we might want to know if it wasn't available (even though that should _never_ be the case for new channels). Refactored it to be an `OptionalRecordT`.

yyforyongyu replied:
yeah keeping it optional would be nice since revocation log can be huge, so even a default value would take up quite some space.

guggero commented on lnwallet/commitment.go:
```diff
@@ -680,15 +760,89 @@ type unsignedCommitmentTx struct {
+		return fn.None[CommitAuxLeaves](), nil
+	}
+
+	return leafStore.UnsafeFromSome().FetchLeavesFromCommit(
+		chanState, commit, keyRing,
+	)
```
Refactored this. Makes sense, now that I understand the whole library a bit better.

guggero commented on lnwallet/commitment.go:
```diff
@@ -610,30 +612,108 @@ func CommitScriptAnchors(chanType channeldb.ChannelType,
+// a state transition function for the blobs to allow them to be refreshed with
+// each state.
+type AuxLeafStore interface {
+	// FetchLeavesFromView attempts to fetch the auxiliary leaves that
+	// correspond to the passed aux blob, and pending original (unfiltered)
+	// HTLC view.
+	FetchLeavesFromView(chanState *channeldb.OpenChannel, prevBlob tlv.Blob,
+		unfilteredView *HtlcView, whoseCommit lntypes.ChannelParty,
+		ourBalance, theirBalance lnwire.MilliSatoshi,
+		keyRing CommitmentKeyRing) (fn.Option[CommitAuxLeaves],
+		CommitSortFunc, error)
```
Added an input and result type and also changed things to return `fn.Result`.

guggero commented on input/taproot.go:
```diff
@@ -31,6 +31,24 @@ func NoneTapLeaf() AuxTapLeaf {
+	SecondLevelLeaf AuxTapLeaf
+}
+
+// AuxTapLeaves is a type alias for a slice of optional tapscript leaves.
+type AuxTapLeaves = map[HtlcIndex]HtlcAuxLeaf
```
Yeah, I noticed the same. Renamed this to `HtlcAuxLeaves`.

guggero commented on channeldb/channel.go:
```diff
@@ -248,6 +248,10 @@ type chanAuxData struct {
+
+	// tapscriptRoot is the optional Tapscript root the channel funding
+	// output commits to.
+	tapscriptRoot tlv.OptionalRecordT[tlv.TlvType6, [32]byte]
```
The extra data is only used for serialization and de-serialization. It houses all the fields that are new and stored in a TLV stream (as opposed to the old `lnwire.WriteElements()` way). If we have the fields in a struct, we can use the new `tlv.RecordT` and `tlv.OptionalRecordT` types, which is nice. But because we don't want to refactor everything to access those records, we still copy from/to the `OpenChannel` struct. I renamed this to `openChannelTlvData` and added a comment to make this a bit more clear.

guggero commented on lnwallet/commitment.go:
```diff
@@ -610,30 +612,108 @@ func CommitScriptAnchors(chanType channeldb.ChannelType,
+	return CommitAuxLeaves{
+		LocalAuxLeaf:       c.RemoteAuxLeaf,
+		RemoteAuxLeaf:      c.LocalAuxLeaf,
+		OutgoingHtlcLeaves: c.IncomingHtlcLeaves,
```
You're right. But this method wasn't used at all, so I removed it.

guggero commented on lnwallet/commitment.go:
```diff
@@ -680,15 +760,89 @@ type unsignedCommitmentTx struct {
+		return fn.None[CommitAuxLeaves](), nil, nil
+	}
+
+	if prevBlob.IsNone() {
```
It shouldn't. But we didn't want to call `UnsafeUnwrap` without at least calling `IsNone()` first. This is mute after the latest refactor though.

guggero commented:
@ProofOfKeags and @yyforyongyu: I believe I have addressed all of your comments to **part 1** and have extracted those commits into https://github.com/lightningnetwork/lnd/pull/9030 as requested.

Please do the second round of review there. I hope the comments here are sufficiently resolved so it's not too much of a hassle to now have a new PR...

guggero commented on lnwire/extra_bytes.go:
```diff
@@ -105,6 +140,45 @@ func (e *ExtraOpaqueData) ExtractRecords(recordProducers ...tlv.RecordProducer)
 	return tlvStream.DecodeWithParsedTypesP2P(extraBytesReader)
 }
 
+// RecordProducers parses ExtraOpaqueData into a slice of TLV record producers
+// by interpreting it as a TLV map.
+func (e *ExtraOpaqueData) RecordProducers() ([]tlv.RecordProducer, error) {
+	var recordProducers []tlv.RecordProducer
+
+	// If the instance is nil or empty, return an empty slice.
+	if e == nil || len(*e) == 0 {
+		return recordProducers, nil
+	}
+
+	// Parse the extra opaque data as a TLV map.
+	tlvMap, err := e.ExtractRecords()
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the TLV map into a slice of records.
+	tlvMapGeneric := make(map[uint64][]byte)
+	for k, v := range tlvMap {
+		tlvMapGeneric[uint64(k)] = v
+	}
+	records := tlv.MapToRecords(tlvMapGeneric)
+
+	// Convert the records to record producers.
+	recordProducers = make([]tlv.RecordProducer, len(records))
+	for i, record := range records {
+		recordProducers[i] = &recordProducer{record}
+	}
+
+	return recordProducers, nil
+}
```
Refactored a bunch of the methods/functions in this file and consolidated some duplication with the `CustomRecords` type.

guggero commented on lnwire/lnwire_test.go:
```diff
@@ -389,6 +389,45 @@ func TestEmptyMessageUnknownType(t *testing.T) {
+	err := customRecords.Validate()
+	require.NoError(t, err)
+
+	// If by chance we end up with an empty set of custom records, we'll
+	// set it to nil to simplify test comparisons.
+	if len(customRecords) == 0 {
+		customRecords = nil
+	}
```
It actually was possible because of the use of `r.Intn(10)`, which can return `0` as a value. Changed it to `r.Intn(9) + 1` instead.

ellemouton commented on invoices/invoiceregistry.go:
```diff
@@ -1027,6 +1032,53 @@ func (i *InvoiceRegistry) notifyExitHopHtlcLocked(
+		// If the interceptor service has provided a response, we'll
+		// use the interceptor session to wait for the client to respond
+		// with a settlement resolution.
+		var interceptErr error
+		interceptSession.WhenSome(func(session InterceptSession) {
+			log.Debug("Waiting for client response from invoice " +
```
can we not hide pretty much all of this within the `HTLCModifier`'s `Intercept call instead ? 
Since I think this will look roughly the same no matter where this is being called from. ie, i think this can then just be: 

```
               err := i.cfg.HtlcModifier.Intercept(
			clientReq, func(resp HtlcModifyResponse) func() {
				if resp.AmountPaid != 0 {
					ctx.amtPaid = resp.AmountPaid
				}
			}
		)
		if err != nil {
			return nil, fmt.Errorf("error during invoice HTLC "+
				"interception: %w", err)
		}
```

ellemouton replied:
the various session channels can then be unexported and the waiting no them can happen within `Intercept` itself

guggero replied:
I like that very much! Was able to simplify things quite a bit.

ellemouton commented on invoices/invoiceregistry.go:
```diff
@@ -1027,6 +1032,53 @@ func (i *InvoiceRegistry) notifyExitHopHtlcLocked(
+
+				interceptErr = err
+
+			case <-session.Quit:
+				// At this point, the interceptor session has
+				// quit.
```
do we want `updateInvoice` to continue if the interceptor session has quit?

ellemouton commented on lnrpc/invoicesrpc/htlc_modifier.go:
```diff
@@ -0,0 +1,110 @@
+// corresponding responses.
+func (r *htlcModifier) run() error {
+	// Register our invoice modifier.
+	r.cfg.modificationInterceptor.SetClientCallback(r.onIntercept)
+	defer r.cfg.modificationInterceptor.SetClientCallback(nil)
+
```
following on from my previous comment, this could perhaps instead be:

```
unregister := r.cfg.modificationInterceptor.SetClientCallback(r.onIntercept)
defer unregister()
```

and perhaps `RegisterInterceptor` is maybe a more suitable name than `SetClientCallback`?

ellemouton commented on lnrpc/invoicesrpc/invoices_server.go:
```diff
@@ -446,3 +464,27 @@ func (s *Server) LookupInvoiceV2(ctx context.Context,
+	modifierServer Invoices_HtlcModifierServer) error {
+
+	// Ensure that there is only one invoice HTLC modifier RPC server
+	// instance.
+	if !s.invoiceHtlcModifierActive.CompareAndSwap(0, 1) {
+		return ErrInvoiceHtlcModifierAlreadyExists
+	}
+	defer s.invoiceHtlcModifierActive.CompareAndSwap(1, 0)
```
follow on from previous comment: i think this is basically redundant since we can just check in the modifier if `callback` is nil or not & ensure that if it is not nil, then no new interceptor can be registered

ellemouton commented on itest/lnd_invoice_acceptor_test.go:
```diff
@@ -0,0 +1,241 @@
+	ht.AssertTopologyChannelOpen(alice, cpBC)
+
+	// Initiate Carol's invoice HTLC modifier.
+	invoiceModifier, cancelModifier := carol.RPC.InvoiceHtlcModifier()
```
maybe can also cover the case of calling this twice and asserting that that errors due to a modifier already existing

ellemouton reviewed (commented):
First pass of all the Invoice HTLC modifier commits in Part 4 :) 
Will do second half of part 4 next :)

ellemouton commented on lnwallet/channel.go:
```diff
@@ -8665,17 +8710,101 @@ func CreateCooperativeCloseTx(fundingTxIn wire.TxIn,
+					extraTxOut.Value,
+				)
+				if amtAfterOutput <= localDust {
+					txOut.PkScript = extraTxOut.PkScript
```
should we not also add the value of the extra output to the txOut here since we no longer will create that output?

ellemouton replied:
ah ok from the subtraction below, I guess this is by default already the case that the value is included in the main output.

ellemouton commented on lnwallet/channel.go:
```diff
@@ -8665,17 +8710,101 @@ func CreateCooperativeCloseTx(fundingTxIn wire.TxIn,
+					return
+				}
+
+				txOut.Value -= extraTxOut.Value
```
do we check somewhere that the extraTxOut's value is in fact <= txOut.Value?

also, do we check somewhere that the new txOut.Value here is above our dust or are we assuming that the check above captures us & ie that ourBalance is equal to txOut.Value?

guggero replied:
Yes, since `txOut.Value` in this case is equal to the local balance from the `amtAfterOutput` subtraction above.
I renamed things to make this more clear.

ellemouton reviewed (commented):
first pass of part 4 chan closer section 👍  looking gooood!

guggero commented on aliasmgr/aliasmgr.go:
```diff
@@ -521,6 +538,18 @@ func (m *Manager) RequestAlias() (lnwire.ShortChannelID, error) {
+			// This is to avoid collision with custom added SCID
+			// aliases that fall into the same range.
+			for {
+				if !haveNextAlias(nextAlias) {
```
Yes, it's possible. The DB here just stores the last monotonically increasing alias we give out when requested through this. But when we add custom aliases (derived from a hash, so more or less randomly distributed across the allowed range), these might collide with the monotonically increasing ones. That doesn't mean the custom ones aren't also on disk somewhere, just not in the bucket we're looking at here.

guggero replied:
Added a couple of sentences more to the comment to make this more clear.

guggero commented on aliasmgr/aliasmgr.go:
```diff
@@ -535,6 +564,18 @@ func (m *Manager) RequestAlias() (lnwire.ShortChannelID, error) {
+				break
+			}
+
+			nextAlias = getNextScid(nextAlias)
```
Added.

guggero commented on routing/bandwidth.go:
```diff
@@ -18,7 +22,40 @@ type bandwidthHints interface {
+	// ProduceHtlcExtraData is a function that, based on the previous extra
+	// data blob of an HTLC, may produce a different blob or modify the
+	// amount of bitcoin this htlc should carry.
+	ProduceHtlcExtraData(totalAmount lnwire.MilliSatoshi,
+		htlcCustomRecords lnwire.CustomRecords) (lnwire.MilliSatoshi,
+		tlv.Blob, error)
```
You're right. In any other case we'll just store the data that's returned from the aux component as an opaque blob of data. But here we actually parse the records again to insert them into the HTLC message. So `lnd` does make assumptions about the content of the blob, so we might as well just return the underlying map and save ourselves an encode/decode cycle.

bitromortac reviewed (commented):
Great work @ all involved with the second part! We may need to persist first hop custom records, to be able to restart payments. Also have some nits and restructuring suggestions (mostly non-blocking).

guggero commented on itest/lnd_forward_interceptor_test.go:
```diff
@@ -344,6 +345,143 @@ func testForwardInterceptorBasic(ht *lntest.HarnessTest) {
+
+	// Resume the intercepted HTLC with a modified amount and custom
+	// records.
+	if packet.CustomRecords == nil {
```
Yeah, that was a bit weird. Should be always `nil` here. Refactored.

guggero commented on lnwire/update_fulfill_htlc.go:
```diff
@@ -23,6 +26,10 @@ type UpdateFulfillHTLC struct {
 
+	// CustomRecords maps TLV types to byte slices, storing arbitrary data
+	// intended for inclusion in the ExtraData field.
+	CustomRecords CustomRecords
```
Yes, you are correct, these are currently not used in any asset use cases and just added for symmetry/completeness.

guggero commented on lnwire/update_fulfill_htlc_test.go:
```diff
@@ -0,0 +1,148 @@
+	// Define test cases.
+	testCases := make([]testCaseUpdateFulfill, 0)
+
+	testCases = append(testCases, testCaseUpdateFulfill{
```
I added a bunch of unit tests in the latest refactor.

guggero commented on lnwire/lnwire_test.go:
```diff
@@ -1430,6 +1430,21 @@ func TestLightningWireProtocol(t *testing.T) {
+			_, err = r.Read(req.PaymentPreimage[:])
+			require.NoError(t, err)
+
+			req.CustomRecords = randCustomRecords(t, r)
```
Not necessarily. The custom records are merged/serialized **into** the `ExtraData` field when going on the wire. So indirectly we are setting the field.

guggero commented on lnrpc/routerrpc/forward_interceptor.go:
```diff
@@ -123,6 +123,13 @@ func (r *forwardInterceptor) resolveFromClient(
 	case ResolveHoldForwardAction_RESUME_MODIFIED:
 		// Modify HTLC and resume forward.
+		incomingAmtMsat := fn.None[lnwire.MilliSatoshi]()
+		if in.IncomingAmountMsat > 0 {
```
AFAIK "settle" is just "here's the preimage", no additional validation logic. So the incoming amount for the validation logic only needs to be set for the adds.

guggero commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -1531,6 +1544,138 @@ func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
+				// Trying to add an alias that we already have
+				// for another channel is wrong.
+				if exists && base != baseScid {
+					return true, fmt.Errorf("%w: alias %v "+
```
I guess it doesn't matter since in the error case we won't look at the boolean value. But to be very precise, in this error case the alias _does_ exist, so `true` is correct.

guggero commented on itest/lnd_routing_test.go:
```diff
@@ -746,6 +747,207 @@ func testInvoiceRoutingHints(ht *lntest.HarnessTest) {
+		"--protocol.anchors",
+	}
+
+	carol := ht.NewNode("Carol", scidAliasArgs)
```
Yeah, good idea. Done.

guggero commented on lnrpc/routerrpc/router.proto:
```diff
@@ -1028,9 +1040,9 @@ message ForwardHtlcInterceptResponse {
-    map<uint64, bytes> custom_records = 8;
+    // Outgoing htlc wire custom records is used to set the p2p message custom
+    // records field for resuming a HTLC.
+    map<uint64, bytes> outgoing_htlc_wire_custom_records = 8;
```
Yeah, you're right. I've streamlined the commits and also the naming. Settled on `InWireCustomRecords` and `OutCustomRecords` where direction is important and just `CustomRecords` where it's kind of clear where they are coming from.

guggero commented on htlcswitch/interceptable_switch.go:
```diff
@@ -641,6 +660,74 @@ func (f *interceptedForward) Resume() error {
+				}
+
+				// Type cast and validate custom records.
+				htlc.CustomRecords = lnwire.CustomRecords(
```
No, I don't think that would be correct. The `f.packet.customRecords` explicitly mentions "payload", which I understand to be the custom records from the onion. So we only set this on the `lnwire` HTLC, as there it's kind of clear it's only what goes on the wire (and not the onion payload).

guggero commented on contractcourt/breach_arbitrator.go:
```diff
@@ -1466,6 +1526,15 @@ func (b *BreachArbitrator) createSweepTx(inputs ...input.Input) (*wire.MsgTx,
+	hasBlobs := fn.Any(func(i input.Input) bool {
+		return i.ResolutionBlob().IsSome()
+	}, inputs)
+	if hasBlobs {
```
The goal is to extract the assets that are swept into their own output (separate assets from BTC). So there's only _one_ additional output necessary. Updated the comment to be more clear.

Crypt-iQ replied:
If we have to split out all the HTLC outputs individually due to our sweep-all justice txn not confirming in time, that will mean an equal number of extra outputs. Just bringing it up so that you're aware.

Roasbeef replied:
Yeah that will be covered in an upcoming portion @Crypt-iQ. As is, this only works with the default sweep scenario where we just need to sweep the commitment output.

guggero commented on invoices/settlement_interceptor.go:
```diff
@@ -0,0 +1,224 @@
+
+		// By this point, we've already checked that the client callback
+		// is set. However, if the client callback has been set to nil
+		// since that check then Exec will return an error.
```
Was left over from a previous iteration where there was a safe callback. Removed it because I didn't understand it back then, but brought it back because it seems to be needed.

guggero commented on lnrpc/invoicesrpc/invoices.proto:
```diff
@@ -55,10 +55,19 @@ service Invoices {
+    server will send HTLCs of invoices to the client and the client can modify
+    some aspects of the HTLC in order to pass the invoice acceptance tests.
+    */
+    rpc HtlcModifier (stream HtlcModifyResponse)
```
Hmm, could perhaps be useful in the future. Not sure if we need to optimize prematurely. Since this will be called way less often than the forwarding HTLC interceptor for example.

guggero commented:
@ellemouton (and @bhandras), I addressed all your comments for part 4.

ellemouton commented on invoices/invoiceregistry.go:
```diff
@@ -1027,6 +1032,53 @@ func (i *InvoiceRegistry) notifyExitHopHtlcLocked(
+			CurrentHeight:      uint32(ctx.currentHeight),
+			Invoice:            *inv,
+		}
+		interceptSession := i.cfg.HtlcModifier.Intercept(
```
cool - looks resolved in the split out PR 🙏

ellemouton commented on lnrpc/invoicesrpc/invoices.proto:
```diff
@@ -55,10 +55,19 @@ service Invoices {
+    intercept and modify the HTLCs that attempt to settle the given invoice. The
+    server will send HTLCs of invoices to the client and the client can modify
+    some aspects of the HTLC in order to pass the invoice acceptance tests.
+    */
+    rpc HtlcModifier (stream HtlcModifyResponse)
```
gotcha - sounds good 👍

yyforyongyu commented on sweep/sweeper.go:
```diff
@@ -1547,8 +1551,21 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
 	// transactions like funding of a channel.
 	sweepWithLock := func(set InputSet) error {
 		return s.cfg.Wallet.WithCoinSelectLock(func() error {
+			extraBudgetRes := fn.MapOptionZ(
```
```diff
diff --git a/server.go b/server.go
index b99e9be44..e97917be2 100644
--- a/server.go
+++ b/server.go
@@ -1125,7 +1125,7 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	}
 
 	aggregator := sweep.NewBudgetAggregator(
-		cc.FeeEstimator, sweep.DefaultMaxInputsPerTx,
+		cc.FeeEstimator, sweep.DefaultMaxInputsPerTx, implCfg.AuxSweeper,
 	)
 
 	s.txPublisher = sweep.NewTxPublisher(sweep.TxPublisherConfig{
@@ -1151,7 +1151,6 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 		Aggregator:           aggregator,
 		Publisher:            s.txPublisher,
 		NoDeadlineConfTarget: cfg.Sweeper.NoDeadlineConfTarget,
-		AuxSweeper:           implCfg.AuxSweeper,
 	})
 
 	s.utxoNursery = contractcourt.NewUtxoNursery(&contractcourt.NurseryConfig{
diff --git a/sweep/aggregator.go b/sweep/aggregator.go
index 6028adca3..71c9baa97 100644
--- a/sweep/aggregator.go
+++ b/sweep/aggregator.go
@@ -5,6 +5,7 @@ import (
 
 	"github.com/btcsuite/btcd/btcutil"
 	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/fn"
 	"github.com/lightningnetwork/lnd/input"
 	"github.com/lightningnetwork/lnd/lntypes"
 	"github.com/lightningnetwork/lnd/lnwallet"
@@ -31,6 +32,10 @@ type BudgetAggregator struct {
 	// maxInputs specifies the maximum number of inputs allowed in a single
 	// sweep tx.
 	maxInputs uint32
+
+	// AuxSweeper is an optional interface that can be used to modify the
+	// way sweep transaction are generated.
+	auxSweeper fn.Option[AuxSweeper]
 }
 
 // Compile-time constraint to ensure BudgetAggregator implements UtxoAggregator.
@@ -38,11 +43,12 @@ var _ UtxoAggregator = (*BudgetAggregator)(nil)
 
 // NewBudgetAggregator creates a new instance of a BudgetAggregator.
 func NewBudgetAggregator(estimator chainfee.Estimator,
-	maxInputs uint32) *BudgetAggregator {
+	maxInputs uint32, auxSweeper fn.Option[AuxSweeper]) *BudgetAggregator {
 
 	return &BudgetAggregator{
-		estimator: estimator,
-		maxInputs: maxInputs,
+		estimator:  estimator,
+		maxInputs:  maxInputs,
+		auxSweeper: auxSweeper,
 	}
 }
 
@@ -159,7 +165,7 @@ func (b *BudgetAggregator) createInputSets(inputs []SweeperInput,
 
 		// Create an InputSet using the max allowed number of inputs.
 		set, err := NewBudgetInputSet(
-			currentInputs, deadlineHeight,
+			currentInputs, deadlineHeight, b.auxSweeper,
 		)
 		if err != nil {
 			log.Errorf("unable to create input set: %v", err)
@@ -173,7 +179,7 @@ func (b *BudgetAggregator) createInputSets(inputs []SweeperInput,
 	// Create an InputSet from the remaining inputs.
 	if len(remainingInputs) > 0 {
 		set, err := NewBudgetInputSet(
-			remainingInputs, deadlineHeight,
+			remainingInputs, deadlineHeight, b.auxSweeper,
 		)
 		if err != nil {
 			log.Errorf("unable to create input set: %v", err)
diff --git a/sweep/mock_test.go b/sweep/mock_test.go
index f999a4aa0..cd5db358a 100644
--- a/sweep/mock_test.go
+++ b/sweep/mock_test.go
@@ -233,9 +233,7 @@ func (m *MockInputSet) FeeRate() chainfee.SatPerKWeight {
 // AddWalletInputs adds wallet inputs to the set until a non-dust
 // change output can be made. Return an error if there are not enough
 // wallet inputs.
-func (m *MockInputSet) AddWalletInputs(wallet Wallet,
-	_ btcutil.Amount) error {
-
+func (m *MockInputSet) AddWalletInputs(wallet Wallet) error {
 	args := m.Called(wallet)
 
 	return args.Error(0)
@@ -243,7 +241,7 @@ func (m *MockInputSet) AddWalletInputs(wallet Wallet,
 
 // NeedWalletInput returns true if the input set needs more wallet
 // inputs.
-func (m *MockInputSet) NeedWalletInput(_ btcutil.Amount) bool {
+func (m *MockInputSet) NeedWalletInput() bool {
 	args := m.Called()
 
 	return args.Bool(0)
diff --git a/sweep/sweeper.go b/sweep/sweeper.go
index 0b80fe904..6257faac1 100644
--- a/sweep/sweeper.go
+++ b/sweep/sweeper.go
@@ -362,10 +362,6 @@ type UtxoSweeperConfig struct {
 	// NoDeadlineConfTarget is the conf target to use when sweeping
 	// non-time-sensitive outputs.
 	NoDeadlineConfTarget uint32
-
-	// AuxSweeper is an optional interface that can be used to modify the
-	// way sweep transaction are generated.
-	AuxSweeper fn.Option[AuxSweeper]
 }
 
 // Result is the struct that is pushed through the result channel. Callers can
@@ -1551,21 +1547,8 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
 	// transactions like funding of a channel.
 	sweepWithLock := func(set InputSet) error {
 		return s.cfg.Wallet.WithCoinSelectLock(func() error {
-			extraBudgetRes := fn.MapOptionZ(
-				s.cfg.AuxSweeper,
-				func(s AuxSweeper) fn.Result[btcutil.Amount] {
-					return s.ExtraBudgetForInputs(
-						set.Inputs(),
-					)
-				},
-			)
-			extraBudget, err := extraBudgetRes.Unpack()
-			if err != nil {
-				return err
-			}
-
 			// Try to add inputs from our wallet.
-			err = set.AddWalletInputs(s.cfg.Wallet, extraBudget)
+			err := set.AddWalletInputs(s.cfg.Wallet)
 			if err != nil {
 				return err
 			}
@@ -1581,21 +1564,8 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
 	}
 
 	for _, set := range sets {
-		extraBudgetRes := fn.MapOptionZ(
-			s.cfg.AuxSweeper,
-			func(s AuxSweeper) fn.Result[btcutil.Amount] {
-				return s.ExtraBudgetForInputs(
-					set.Inputs(),
-				)
-			},
-		)
-		extraBudget, err := extraBudgetRes.Unpack()
-		if err != nil {
-			log.Errorf("Failed to get extra budget for set %v: %v",
-				set, err)
-		}
-
-		if set.NeedWalletInput(extraBudget) {
+		var err error
+		if set.NeedWalletInput() {
 			// Sweep the set of inputs that need the wallet inputs.
 			err = sweepWithLock(set)
 		} else {
diff --git a/sweep/tx_input_set.go b/sweep/tx_input_set.go
index d26d3d75e..3777a8a37 100644
--- a/sweep/tx_input_set.go
+++ b/sweep/tx_input_set.go
@@ -37,11 +37,11 @@ type InputSet interface {
 	// AddWalletInputs adds wallet inputs to the set until a non-dust
 	// change output can be made. Return an error if there are not enough
 	// wallet inputs.
-	AddWalletInputs(wallet Wallet, extraBudget btcutil.Amount) error
+	AddWalletInputs(wallet Wallet) error
 
 	// NeedWalletInput returns true if the input set needs more wallet
 	// inputs.
-	NeedWalletInput(extraBudget btcutil.Amount) bool
+	NeedWalletInput() bool
 
 	// DeadlineHeight returns an absolute block height to express the
 	// time-sensitivity of the input set. The outputs from a force close tx
@@ -111,6 +111,12 @@ type BudgetInputSet struct {
 	// deadlineHeight is the height which the inputs in this set must be
 	// confirmed by.
 	deadlineHeight int32
+
+	// extraBudget is a value that should be allocated to sweep the given
+	// set of inputs. This can be used to add extra funds to the sweep
+	// transaction, for example to cover fees for additional outputs of
+	// custom channels.
+	extraBudget btcutil.Amount
 }
 
 // Compile-time constraint to ensure budgetInputSet implements InputSet.
@@ -167,8 +173,8 @@ func validateInputs(inputs []SweeperInput, deadlineHeight int32) error {
 }
 
 // NewBudgetInputSet creates a new BudgetInputSet.
-func NewBudgetInputSet(inputs []SweeperInput,
-	deadlineHeight int32) (*BudgetInputSet, error) {
+func NewBudgetInputSet(inputs []SweeperInput, deadlineHeight int32,
+	auxSweeper fn.Option[AuxSweeper]) (*BudgetInputSet, error) {
 
 	// Validate the supplied inputs.
 	if err := validateInputs(inputs, deadlineHeight); err != nil {
@@ -186,9 +192,31 @@ func NewBudgetInputSet(inputs []SweeperInput,
 
 	log.Tracef("Created %v", bi.String())
 
+	// Optionally attach an extra budget to the input set.
+	auxSweeper.WhenSome(func(aux AuxSweeper) {
+		bi.attachExtraBudget(aux)
+	})
+
 	return bi, nil
 }
 
+// attachExtraBudget...
+func (b *BudgetInputSet) attachExtraBudget(s AuxSweeper) error {
+	extraBudgetRes := s.ExtraBudgetForInputs(b.Inputs())
+
+	extraBudget, err := extraBudgetRes.Unpack()
+	if err != nil {
+		log.Errorf("Failed to get extra budget for set %v: %v",
+			b.Inputs(), err)
+
+		return err
+	}
+
+	b.extraBudget = extraBudget
+
+	return nil
+}
+
 // String returns a human-readable description of the input set.
 func (b *BudgetInputSet) String() string {
 	inputsDesc := ""
@@ -209,13 +237,13 @@ func (b *BudgetInputSet) addInput(input SweeperInput) {
 //
 // A set may need wallet inputs when it has a required output or its total
 // value cannot cover its total budget.
-func (b *BudgetInputSet) NeedWalletInput(extraBudget btcutil.Amount) bool {
+func (b *BudgetInputSet) NeedWalletInput() bool {
 	var (
 		// budgetNeeded is the amount that needs to be covered from
 		// other inputs. We start at the value of the extra budget,
 		// which might be needed for custom channels that add extra
 		// outputs.
-		budgetNeeded = extraBudget
+		budgetNeeded = b.extraBudget
 
 		// budgetBorrowable is the amount that can be borrowed from
 		// other inputs.
@@ -276,9 +304,7 @@ func (b *BudgetInputSet) copyInputs() []*SweeperInput {
 // set to its initial state by removing any wallet inputs added.
 //
 // NOTE: must be called with the wallet lock held via `WithCoinSelectLock`.
-func (b *BudgetInputSet) AddWalletInputs(wallet Wallet,
-	extraBudget btcutil.Amount) error {
-
+func (b *BudgetInputSet) AddWalletInputs(wallet Wallet) error {
 	// Retrieve wallet utxos. Only consider confirmed utxos to prevent
 	// problems around RBF rules for unconfirmed inputs. This currently
 	// ignores the configured coin selection strategy.
@@ -322,7 +348,7 @@ func (b *BudgetInputSet) AddWalletInputs(wallet Wallet,
 			pi.OutPoint(), utxo.Value)
 
 		// Return if we've reached the minimum output amount.
-		if !b.NeedWalletInput(extraBudget) {
+		if !b.NeedWalletInput() {
 			return nil
 		}
 	}
```

guggero replied:
Ah, much better! Thank you very much :pray:

yyforyongyu commented on contractcourt/breach_arbitrator.go:
```diff
@@ -1539,19 +1596,43 @@ func (b *BreachArbitrator) sweepSpendableOutputsTxn(txWeight lntypes.WeightUnit,
+	// amount from the sweep amt.
+	if b.cfg.AuxSweeper.IsSome() {
+		extraChangeOut.WhenResult(func(o sweep.SweepOutput) {
+			sweepAmt -= o.Value
```
is it possible that `o.Value` is greater than or equal to `sweepAmt`?

Crypt-iQ replied:
imo we should be defensive here in case the call to `DeriveSweepAddr` returns something funky

Roasbeef replied:
So with the way things work as is, the current o.Value will always be a static amount that's above dust, and we won't try to sweep anything that is around that dust level. 

With that said, I think it makes sense to add a more defensive check here.

yyforyongyu reviewed (changes requested):
Getting close! The CI needs to be fixed, in addition left a few questions and suggestions.

Crypt-iQ commented on lnwallet/channel.go:
```diff
@@ -2127,6 +2155,38 @@ func NewBreachRetribution(chanState *channeldb.OpenChannel, stateNum uint64,
+			CommitTx:       spendTx,
+			SignDesc:       *br.LocalOutputSignDesc,
+			KeyRing:        keyRing,
+			CsvDelay:       ourDelay,
```
I think by the description of `BreachCsvDelay`: 
```
// BreachCsvDelay is the CSV delay for the remote output. This is only
// set when the CloseType is Breach. This indicates the CSV delay to
// use for the remote party's to_local delayed output, that is now
// rightfully ours in a breach situation.
```

it is correct as-is

guggero replied:
I also think this is correct because I remember Laolu fixing this with a follow-up commit at some point, so I think we have an itest for it. Can you please confirm @Roasbeef?

Roasbeef replied:
So we need two CSV delays here:
  * One which in most cases is just 1, due the CPFP carve out. 
  * When handling breaches, we need to reconstruct the script of the remote party, which will use the CSV of their actual commitment transaction.

Crypt-iQ commented on contractcourt/breach_arbitrator.go:
```diff
@@ -1636,12 +1639,28 @@ func taprootBriefcaseFromRetInfo(retInfo *retributionInfo) *taprootBriefcase {
+				)
+			})
+
 		// For spending the HTLC outputs, we'll store the first and
```
I think we need to store the `resolutionBlob` for htlc's for the restart case? Also since I'm unaware of the custom channels commitment format, does the blob change if the htlc is brought to the second level?

Roasbeef replied:
The blob doesn't change if the HTLC is brought to the second level. We can essentially pre-sign for it all. I have another PR that'll fill in some of the HTLC gaps here coming up.

Crypt-iQ reviewed (commented):
Left some comments. 

Two questions:
- do we need to create new witness sizes in `input/size.go`?
- do we need to worry about the `Aux*` things blocking?

guggero commented on lnwallet/channel.go:
```diff
@@ -2127,6 +2155,38 @@ func NewBreachRetribution(chanState *channeldb.OpenChannel, stateNum uint64,
+
+		// At this point, we'll check to see if we need any extra
+		// resolution data for this output.
+		resolveReq := ResolutionReq{
```
Not sure I understand what you mean? The `if` block is done at this point, so this would only not be executed if an error occurred (or `ourAmt < int64(chanState.RemoteChanCfg.DustLimit)` from the outer condition).

Roasbeef replied:
Yeah this is the case where we've already confirmed on chain, and want to create the resolution blob needed to sweep the output into our wallet.

guggero commented on lnwallet/channel.go:
```diff
@@ -2127,6 +2155,38 @@ func NewBreachRetribution(chanState *channeldb.OpenChannel, stateNum uint64,
+			ShortChanID:    chanState.ShortChanID(),
+			Initiator:      chanState.IsInitiator,
+			FundingBlob:    chanState.CustomBlob,
+			Type:           input.TaprootRemoteCommitSpend,
```
I think this is correct. Since we're in the `ourAmt >= dust` branch, this is the remote commitment's unencumbered `to_remote` output that goes to us. We have the `TaprootCommitmentRevoke` in the `theirAmt >= dust` branch below.
@Roasbeef can you confirm please?

Roasbeef replied:
Yeah this is fine here, on the server side we're able to recognize the breach case properly.

guggero commented on lnwallet/channel.go:
```diff
@@ -2174,6 +2234,35 @@ func NewBreachRetribution(chanState *channeldb.OpenChannel, stateNum uint64,
+					CommitTx:       spendTx,
+					SignDesc:       *br.RemoteOutputSignDesc,
+					KeyRing:        keyRing,
+					CsvDelay:       theirDelay,
```
We have the [normal force close test](https://github.com/lightninglabs/lightning-terminal/blob/0-19-staging/itest/litd_custom_channels_test.go#L1107) and a [breach test](https://github.com/lightninglabs/lightning-terminal/blob/0-19-staging/itest/litd_custom_channels_test.go#L1465), yes.

guggero commented on lnwallet/channel.go:
```diff
@@ -6527,10 +6621,14 @@ type UnilateralCloseSummary struct {
 // which case we will attempt to sweep the non-HTLC output using the passed
 // commitPoint.
+//
+//nolint:funlen
```
Yeah, I know... Did some light refactor to get us to exactly 200 lines of code :sweat_smile:

guggero commented on sweep/fee_bumper.go:
```diff
@@ -1268,31 +1271,69 @@ func (t *TxPublisher) createSweepTx(inputs []input.Input,
+	extraTxOut := fn.MapOption(
+		func(sweepOuts []SweepOutput) fn.Option[SweepOutput] {
+			for _, sweepOut := range sweepOuts {
+				if sweepOut.IsExtra {
```
Correct, there's only ever going to be a single extra output when sweeping channel outputs. Assets from multiple swept inputs are merged into a single extra output.

guggero commented on sweep/fee_bumper.go:
```diff
@@ -1268,31 +1271,69 @@ func (t *TxPublisher) createSweepTx(inputs []input.Input,
+	// we'll attempt to see if we have any other change outputs we'll need
+	// to add to the sweep transaction.
+	changePkScripts := [][]byte{changePkScript.DeliveryAddress}
+	extraChangeOut := fn.MapOptionZ(
```
You're right, this is a bit weird. Refactored it slightly. Hope it looks better now.

guggero commented:
I didn't get to addressing all comments before my break. I resolved those that are addressed.

Roasbeef commented:
> do we need to create new witness sizes in input/size.go?


No, nothing changes re the witness size.

Roasbeef commented on sweep/fee_bumper.go:
```diff
@@ -1379,12 +1437,16 @@ func prepareSweepTx(inputs []input.Input, changePkScript []byte,
 	// Otherwise, we'll actually recognize it as a change output.
 	default:
-		// TODO(roasbeef): Implement (later commit in this PR).
+		changeOuts = append(changeOuts, SweepOutput{
```
So typically we like to keep the TAP assets and the normal Bitcoin in distinct outputs. If we combined the TAP output and the normal change output, then lnd wouldn't be able to move the funds without contacting tapd.

Roasbeef commented on sweep/fee_bumper.go:
```diff
@@ -131,6 +132,8 @@ type BumpRequest struct {
 	// Get the size of the sweep tx, which will be used to calculate the
 	// budget fee rate.
+	//
+	// TODO(roasbeef): also wants the extra change output?
```
Fixed in the latest push.

Roasbeef commented on sweep/sweeper.go:
```diff
@@ -1564,8 +1581,21 @@ func (s *UtxoSweeper) sweepPendingInputs(inputs InputsMap) {
+				set, err)
+		}
+
+		if set.NeedWalletInput(extraBudget) {
```
So this area has changed. and now the budget is computed earlier, without any additional duplication.

yyforyongyu reviewed (commented):
Cool think all my comments are addressed - just need to fix the merge conflict to let the CI run.

Roasbeef commented:
> Cool think all my comments are addressed - just need to fix the merge conflict to let the CI run.

Ok, fixed the last merge conflict and also an issue in the way I updated `MaxFeeRateAllowed`. We're green now!
